### PR TITLE
Bytestring bits

### PR DESCRIPTION
**Pull request by jrudolph**
Two performance improvements for ByteString and ByteStringBuilder mostly for 2.13.

    actor: improve ByteStringBuilder.++= / addAll
    
    Especially in the Scala 2.13 version, the previous `if xs.iterator.isEmpty`
    for the common case was a big problem, since it expensively created an
    iterator even for the most common ++=(ByteString) case.

    actor: implement ByteString.isEmpty in as simple terms as possible
    
    It turned up in profiles, the usual implementation in 2.13 is through
    SeqOps.isEmpty -> lengthCompare which checks if knownSize != -1 for
    the fast path. All of that doesn't sound too bad but introduces enough
    indirection that the inliner might not be able to inline all of that
    which then leads to a multimorphic callsite e.g. to call lengthCompare.

### CODE REVIEW

File: akka-actor/src/main/scala-2.12/akka/util/ByteString.scala [hunk 1]
```diff
@@ -1045,37 +1047,45 @@ final class ByteStringBuilder extends Builder[Byte, ByteString] {
...
+          _length += b.length
+        case b: ByteString1 =>
+          clearTemp()
+          _builder += b
+          _length += b.length
+        case bs: ByteStrings =>
+          clearTemp()
+          _builder ++= bs.bytestrings
+          _length += bs.length
+      }
```
Comment (johanandren) [inline akka-actor/src/main/scala-2.12/akka/util/ByteString.scala:1065]: `clearTemp()` happens for all cases so perhaps pull out and do before patmatch?

**Review by johanandren — APPROVED** [review]

**Comment by johanandren** [comment]
Test failure is unrelated (revert PR for that is in progress)

File: akka-actor/src/main/scala-2.12/akka/util/ByteString.scala (hunk 1 above)
Comment (jrudolph) [reply akka-actor/src/main/scala-2.12/akka/util/ByteString.scala:1063]: :+1:, why not. Done

**Review by ennru — APPROVED** [review]
LGTM.

**Pull request merged**